### PR TITLE
Fix role impersonation for Agentless nodes

### LINKED ISSUE #25083: Role impersonated certificates do not work with Agentless SSH proxy re-issuing
Certificates issued by Machine ID do not work with Agentless SSH proxy re-issuing, the roles on the certificate are not considered and the OpenSSH CA cert that is generated is based purely off the username of the bot, this means the logins that are granted by the impersonated roles are not included within the principals of the generated certificate.

I believe there's a chance this will effect access requests too.

### PR DESCRIPTION
**Pull request by capnspacehook**
Fixes https://github.com/gravitational/teleport/issues/25083.

### CODE REVIEW

**Comment by strideynet** [comment]
Tested at [b68471f](https://github.com/gravitational/teleport/pull/25087/commits/b68471f86fe4321a53b8de6b7c18dc8296842b6e) and this fixes my issue as described in #25083

File: lib/agentless/agentless.go [hunk 1]
```diff
@@ -57,14 +68,14 @@ func SignerFromAuthzContext(authzCtx *authz.Context, generator CertGenerator) fu
...
-		signer, err := createAuthSigner(ctx, authzCtx.User.GetName(), identity.TeleportCluster, ttl, generator)
+		signer, err := createAuthSigner(ctx, authzCtx.User.GetName(), identity.TeleportCluster, ttl, identity.Groups, identity.Traits, generator)
 		return signer, trace.Wrap(err)
 	}
 }
 
 // createAuthSigner creates a [ssh.Signer] that is signed with
 // OpenSSH CA and can be used to authenticate to agentless nodes.
-func createAuthSigner(ctx context.Context, username, clusterName string, ttl time.Duration, generator CertGenerator) (ssh.Signer, error) {
+func createAuthSigner(ctx context.Context, username, clusterName string, ttl time.Duration, roles []string, traits wrappers.Traits, generator CertGenerator) (ssh.Signer, error) {
```
Comment (probakowski) [inline lib/agentless/agentless.go:78]: I'd consider creating struct to pass arguments here as list gets pretty long

**Review by probakowski — APPROVED** [review]
Looks good

**Comment by capnspacehook** [comment]
Forgot to mention, would like to make sure I'm not missing any required checks when generating the certificate signed by the OpenSSH CA. I noticed `lib/auth.(ServerWithRoles).generateUserCerts ` has a ton of checks, but I *think* most of those are unnecessary here, as I am signing a certificate *after* the user has been authenticated and provided a valid certificate signed by the User CA. But I'd like to be sure.

File: lib/agentless/agentless.go (hunk 1 above)
Comment (capnspacehook) [reply lib/agentless/agentless.go:78]: I'll probably add one if it gets any longer, I think it's fine for now as it's not being used anywhere else

**Comment by Joerger** [comment]
> Forgot to mention, would like to make sure I'm not missing any required checks when generating the certificate signed by the OpenSSH CA. I noticed [`lib/auth.(ServerWithRoles).generateUserCerts`](https://github.com/gravitational/teleport/blob/06a3864fc33e8db970afb82d1a17f6b53856121a/lib/auth/auth_with_roles.go#L2669) does a ton of checks before calling `generateUserCert`, but I _think_ most of those are unnecessary here, as I am signing a certificate _after_ the user has been authenticated and provided a valid certificate signed by the User CA. But I'd like to be sure.

I think you do need to carry over the checks in `lib/auth.(ServerWithRoles).generateUserCerts`, as these determine what certificate re-issuing privileges an authenticated user has. The `GenerateUserCerts rpc` also goes through this path for certificate reissuing. This is separate from the `(Server).generateUserCert` checks which are mostly validating and consolidating certificate generation parameters.

IIUC with the current setup it may be possible to generate an OpenSSH certificate for an impersonated user, even though the user's underlying certificate is already an impersonated user.

File: api/proto/teleport/legacy/client/proto/authservice.proto [hunk 2]
```diff
@@ -209,6 +209,12 @@ message OpenSSHCertRequest {
   ];
   // Cluster is the Teleport cluster name.
   string Cluster = 4 [(gogoproto.jsontag) = "cluster"];
+  repeated string Roles = 5 [(gogoproto.jsontag) = "roles"];
```
Comment (Joerger) [inline api/proto/teleport/legacy/client/proto/authservice.proto:212]: Can we just pass the user's TLS certificate instead? This way, the Auth server can parse the user's certificate into a `tlsca.Identity` and use this for the normal certificate issuing flow - https://github.com/gravitational/teleport/blob/06a3864fc33e8db970afb82d1a17f6b53856121a/lib/auth/auth_with_roles.go#L2669. This should circumvent any similar issues with impersonation, access request roles, etc.

**Comment by capnspacehook** [comment]
Working on a fix for what @Joerger brought up in another branch, will close as this branch is not being used anymore.

**Pull request closed without merging**